src/topology/algebra/order/lower.lean

```diff
@@ -0,0 +1,262 @@
+/-
+Copyright (c) 2022 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+
+import topology.basic
+import topology.order
+import topology.separation
+import data.set.intervals.basic
+import order.upper_lower
+
+/-!
+# Lower topology
+
+This file introduces the lower topology on a preorder. It is shown that the lower topology on a
+partial order is T₀ and the non-empty complements of the upper closures of finite subsets form a
+basis.
+
+## Implementation notes
+
+Approach inspired by `order_topology` from topology.algebra.order.basic
+
+## References
+
+* [Gierz et al, A Compendium of Continuous Lattices][GierzEtAl1980]
+
+## Tags
+
+lower topology, preorder
+-/
+
+universes u
+variable {α : Type u}
+
+open  set topological_space
+
+/--
+The lower topology is the topology generated by the complements of the closed intervals to infinity
+-/
+class lower_topology (α : Type u) [t : topological_space α] [preorder α] : Prop :=
+(topology_eq_generate_Ici_comp : t = generate_from {s | ∃a, (Ici a)ᶜ = s })
```

Comment:
```suggestion
(topology_eq_generate_Ici_comp : t = generate_from {s | ∃ a, (Ici a)ᶜ = s})
```
Shouldn't you make `generate_from {s | ∃ a, (Ici a)ᶜ = s}` a definition (which I think should enjoy the name `lower_topology`), considering that it's used below and you'll use it to define the Lawson topology?

And about the design decision: I think it's very rare that you want to assert a pre-existing topology on a type is the lower topology; instead, you want to equip a type with the lower topology. So what I'd do is to introduce a type synonym of `α` (maybe called `lower α` or `with_lower_topology α`, and put the lower topology as a `topological_space` instance. Of course, you also want to transfer other instances you care about (for example the preorder) to this type synonym.

Reply:
@alreadydone There seem to be (at least) two approaches in mathlib. [order_topology](https://github.com/leanprover-community/mathlib/blob/master/src/topology/algebra/order/basic.lean#L688) is defined as a mixin and [Scott](https://github.com/leanprover-community/mathlib/blob/master/src/topology/omega_complete_partial_order.lean#L68) is defined as a type synonym. I wasn't sure which to use to I picked the mixin approach, but I can switch to the type synonym approach if that's preferred?

Reply:
I think [order_topology](https://github.com/leanprover-community/mathlib/blob/master/src/topology/algebra/order/basic.lean?rgh-link-date=2022-10-22T06%3A48%3A29Z#L688) exists because there can be topologies constructed in another way that happen to be the same as the order topology, like [real.order_topology](https://leanprover-community.github.io/mathlib_docs/topology/metric_space/basic.html#real.order_topology); I'm not sure if this is the case for lower topology. And we intend the order topology [preorder.topology](https://leanprover-community.github.io/mathlib_docs/topology/algebra/order/basic.html#preorder.topology) to be the default topology on a preorder, though it's not made an instance due to defeq issues. So we definitely want to have a way of putting the lower topology on a type (or its synonym), but I think for your purpose you don't need a `lower_topology` predicate now (correct me if I'm wrong); if we want it in the future we could just add it easily.

Reply:
@alreadydone Thanks. Should I make the type synonym `@[reducible]` as is done for Scott?

Reply:
@alreadydone I've pushed a commit to use a type synonym. Is that what you meant?

Reply:
Thanks! I think using `@[reducible]` saves you the trouble of copying instances over, but maybe the instances on the type synonym may also leak to the original type? But since `Scott` was done this way, I assume it should be safe or at least workable. @YaelDillies certainly knows better than me on this!

Reply:
I would strongly advise against `@[reducible]`. On the mixin vs lower topology debate, I have little clue, I would need to know how it's used. me just say that they are not mutually exclusive options and ask whether the product of lower topologies is the lower topology (in which case a mixin is probably better so you don't have to transfer everything along `homeomorph (synonym (α × β)) (synonym α × synonym β)`).

Reply:
@YaelDillies Yes, it is. I'm working on that on a separate branch, but haven't completed the proof yet: https://github.com/leanprover-community/mathlib/compare/lower-topology#diff-1eb196fc43d68c866587534cbc7c3e018ed3661efb121a5ad25a33f4ceaba451R379

Reply:
What should the statement "the lower topology on the partial order α × β coincides with the product topology of the lower topology on α with the lower topology on β" now look like?

Reply:
You may construct a homeomorphism between `with_lower_topology (α × β)` and `with_lower_topology α × with_lower_topology β`, or you may state that the [prod.topological_space](https://leanprover-community.github.io/mathlib_docs/topology/constructions.html#prod.topological_space) of the two `lower_topology` equals `lower_topology` on the product.

Reply:
@alreadydone the latter would look something like this?
```
lemma test : with_lower_topology.topological_space (α × β) = @prod.topological_space (with_lower_topology α) (with_lower_topology β) (with_lower_topology.topological_space α) (with_lower_topology.topological_space β) := sorry
```

Reply:
Simpler than that; I meant
```
lemma test : lower_topology (α × β) = @prod.topological_space α β (lower_topology α) (lower_topology β) := sorry
```
For the former, you may use [equiv.prod_congr](https://leanprover-community.github.io/mathlib_docs/logic/equiv/basic.html#equiv.prod_congr) combined with `to_lower`/`of_lower` to construct the `equiv`.

Reply:
@alreadydone Thanks. So the former would look like this?
```
def lower_topology_prod_hom :
  with_lower_topology (α × β) ≃ₜ ((with_lower_topology α) × (with_lower_topology β)) :=
{ continuous_to_fun := sorry,
  continuous_inv_fun := sorry,
  ..(equiv.prod_congr (@with_lower_topology.to_lower α) (@with_lower_topology.to_lower β)) }
```

Reply:
I think the last line should be `of_lower.trans (to_lower.prod_congr to_lower)`.

Reply:
@YaelDillies said on Zulip (in a non-public stream):
> Yeah I don't have much more to say apart from the fact that I don't like the general design (we should have a mixin saying that a topology is the lower topology rather than forcing people to use a specific type)

and that's probably the reason why no one has taken the action to merge this PR. Maybe your further work (including lower topology on products) could shed light on whether the `is_lower_topology` predicate/class is necessary?

Reply:
> @YaelDillies said on Zulip (in a non-public stream):
> 
> > Yeah I don't have much more to say apart from the fact that I don't like the general design (we should have a mixin saying that a topology is the lower topology rather than forcing people to use a specific type)

Isn't that the design I started with?

> 
> and that's probably the reason why no one has taken the action to merge this PR. Maybe your further work (including lower topology on products) could shed light on whether the `is_lower_topology` predicate/class is necessary?

I removed the last `sorry` from my branch with products last night: https://github.com/leanprover-community/mathlib/tree/lower-topology so it seems that the current design works for that purpose. I need to tidy it up a bit, but I'm hoping to turn it into a PR shortly.

Reply:
@alreadydone  If I have hypotheses:
```
s : set (with_lower_topology α),
hs : s ∈ {s : set α | ∃ (a : α), (Ici a)ᶜ = s}
```
How would I cast `s` to type `set α` so that I could `rw mem_set_of_eq at hs,`?

Reply:
Do `of_lower ⁻¹' s`

Reply:
This is the PR regarding the lower topology on products: #17426.